### PR TITLE
Make OCP deployer build its own installer  and client

### PR DESCRIPTION
**Pull request by pebrc**
Part of #4319 

The basic idea is (inspired by the OCP3 deployer) to have separate Docker images with the necessary client programs to stand up the desired K8s cluster within deployer itself. 

This means we move most of the Cloud provider specific bits out the `ci` image and into new Dockefiles that live under `/hack/deployer/clients/$DRIVER_ID` This can be done one by one. 

This PR:
* implements this approach for OCP by using a customer Dockerfile that can be parameterised with the OpenShift version (I think I tested all of the ones we need to run) 
* refactors the existing OCP driver to fix some bugs, remove all the untyped maps and type assertions and hopefully simplify it a bit
* removes the annoying behaviour where the OCP driver would nuke your local kubeconfig, instead I try to merge it with any existing ones (which works as long as there is not already another OCP from deployer one in it) 

Limitations: 
* the OCP driver currently still relies on the gcloud client being present on the host
* no unit tests 

Follow-up PRs:
* set up Jenkins jobs to run the different OCP versions, I am thinking we should maybe have some form of rotation instead of testing all of them all the time. This is mainly to save resources as each of these OCP clusters is 5 VMs + network resources + DNS

### CODE REVIEW

**Comment by pebrc** [comment]
Jenkins test this please

**Review by barkbay — COMMENTED** [review]
Nice work, I'll try to take the time to do some tests but overall I think I'm 👍 with the approach.

File: hack/deployer/runner/client.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	imageExists, err := NewCommand(fmt.Sprintf("docker pull -q %s", image)).OutputContainsAny("Downloading", "Extracting", "Verifying", "complete")
+	return imageExists && err == nil
+}
+
+func clientImageName(driverID, clientVersion, dockerfileName string) (string, error) {
+	// hash Dockerfile to trigger rebuild on content changes
+	f, err := os.Open(dockerfileName)
+	if err != nil {
+		return "", err
+	}
```
[the remark this answers is not in the document]
Comment (pebrc) [reply hack/deployer/runner/client.go:63]: Good catch 🎉

File: hack/deployer/runner/client.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+func checkImageExists(image string) bool {
+	// short circuit if locally available e.g in local dev mode
+	if output, err := NewCommand(fmt.Sprintf("docker images -q %s", image)).Output(); len(output) > 0 && err == nil {
+		return true
+	}
+
+	// check registry
+	imageExists, err := NewCommand(fmt.Sprintf("docker pull -q %s", image)).OutputContainsAny("Downloading", "Extracting", "Verifying", "complete")
```
[the remark this answers is not in the document]
Comment (pebrc) [reply hack/deployer/runner/client.go]: I blindly copied that from our CI Makefile 😞

File: hack/deployer/runner/command.go [hunk 3]
```diff
@@ -52,6 +67,19 @@ func (c *Command) Run() error {
 	return err
 }
 
+func (c *Command) RunWithRetries(numRetries int, timeout time.Duration) error {
```
[the remark this answers is not in the document]
Comment (pebrc) [reply hack/deployer/runner/command.go]: I renamed it.

File: hack/deployer/runner/ocp.go [hunk 4]
```diff
@@ -70,292 +73,389 @@ func init() {
...
+}
+
 type OcpDriver struct {
-	plan Plan
-	ctx  map[string]interface{}
+	plan         Plan
+	runtimeState runtimeState
 }
 
 func (gdf *OcpDriverFactory) Create(plan Plan) (Driver, error) {
```
[the remark this answers is not in the document]
Comment (pebrc) [reply hack/deployer/runner/ocp.go]: The linter tells me to omit the `_` as well, but 👍

**Review by barkbay — APPROVED** [review]
I ran an "happy test" locally and it worked as expected.
I think I'm 👍 with merging it and iterate if needed.

**Pull request merged**